### PR TITLE
[R2] get started guide

### PR DESCRIPTION
**Pull request by deadlypants1973**

### CODE REVIEW

File: content/r2/_index.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+layout: list
+meta:
+  title: Cloudflare R2 documentation
+---
+
+{{<content-column>}}
+
+# Cloudflare R2 (beta) documentation
+
+Cloudflare R2 Storage allows developers to store large amounts of unstructured data without the large egress bandwidth fees associated with typical cloud storage services.
```
Comment (marciocloudflare) [inline content/r2/_index.md]:
```suggestion
Cloudflare R2 Storage allows developers to store large amounts of unstructured data without the costly egress bandwidth fees associated with typical cloud storage services.
```

Comment (marciocloudflare) [reply content/r2/_index.md]: To avoid repetition of the word "large",  my suggestion is you change the second large to "costly"

File: content/r2/_index.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+---
+title: Overview
+type: overview
+pcx-content-type: overview
+weight: 1
+layout: list
```
Comment (marciocloudflare) [inline content/r2/_index.md]: What's this?

File: content/r2/get-started/_index.md [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+weight: 1
+meta:
+  title: Get started guide
+---
+
+{{<content-column>}}
+
+# R2 (beta) get started guide
+
+Cloudflare R2 Storage allows developers to store large amounts of unstructured data without the large egress bandwidth fees associated with typical cloud storage services.
```
Comment (marciocloudflare) [inline content/r2/get-started/_index.md]:
```suggestion
Cloudflare R2 Storage allows developers to store large amounts of unstructured data without the costly egress bandwidth fees associated with typical cloud storage services.
```

Comment (marciocloudflare) [reply content/r2/get-started/_index.md]: Same suggestion to avoid repetition of "large". Also, do yo realize this sentence is the same as the one in the Overview page?

File: content/r2/get-started/_index.md [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+
+{{<Aside type="note" header="Beta registration">}}
+
+As of February 2022, R2 is currently under development. You can [sign up to join the waitlist for access](https://www.cloudflare.com/r2-storage/).
+
+{{</Aside>}}
+
+This guide will instruct you through setting up Wrangler to deploying your first R2 bucket.
```
Comment (marciocloudflare) [inline content/r2/get-started/_index.md]: This structure with 2 gerunds in a row sounds weird to me. Maybe it's my non-native ears, but I'd say that second verb should be "deploy" instead?

File: content/r2/get-started/_index.md [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+
+## 1. Install Wrangler
+
+To create your R2 bucket, install [Wrangler](/get-started/guide#2-install-the-workers-cli), the Workers CLI.
+
+To install [`wrangler`](https://github.com/cloudflare/wrangler), ensure you have [`npm` installed](https://www.npmjs.com/get-npm), preferably using a Node version manager like [Volta](https://volta.sh/) or [nvm](https://github.com/nvm-sh/nvm) to avoid permission issues or to easily change Node.js versions, then run:
```
Comment (marciocloudflare) [inline content/r2/get-started/_index.md]:
```suggestion
To install [`wrangler`](https://github.com/cloudflare/wrangler), ensure you have [`npm` installed](https://www.npmjs.com/get-npm). To avoid permission issues or to easily change Node.js versions, we suggest you use a Node version manager like [Volta](https://volta.sh/) or [nvm](https://github.com/nvm-sh/nvm). Then run:
```

Comment (marciocloudflare) [reply content/r2/get-started/_index.md]: Sentence was too big. Wdyt?

File: content/r2/get-started/_index.md [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+R2 requires a Wrangler version of `1.19.8` or higher. To check your Wrangler version, run `wrangler --version`.
+
+{{</Aside>}}
+
+## 3. Create your bucket
+
+Create your bucket by running:
+
+```sh
+wrangler r2 bucket create <your-bucket-name>
```
Comment (marciocloudflare) [inline content/r2/get-started/_index.md]:
```suggestion
wrangler r2 bucket create <YOUR_BUCKET_NAME>
```

Comment (marciocloudflare) [reply content/r2/get-started/_index.md]: As per PCX style guide conventions

File: content/r2/get-started/_index.md [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+
+After running the `list` command, you will see all bucket names, including the one you have just created.
+
+## 4. Bind your bucket to a Worker
+
+You will need to bind your bucket to a Worker.
+
+{{<Aside type="note" header="Bindings">}}
+
+A binding is a runtime variable that the Workers runtime provides to your code. The capability of these variables can perform a number of operations depending on what they are bound to; for example, a binding may offer the [KV Namespace API](/workers/runtime-apis/kv), the [Durable Object API](/workers/runtime-apis/durable-objects), the [R2 API](#api), or it may be one of your user-defined environment variables. Every binding's variable name and behavior is determined by you when deploying the Worker. Refer to the [Environment Variables](/workers/platform/environment-variables) documentation for more information.
```
Comment (marciocloudflare) [inline content/r2/get-started/_index.md]: I'm not sure what `The capability of these variables can perform a number of operations` means

File: content/r2/get-started/_index.md [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+
+A binding is defined in the `wrangler.toml` file of your Worker project's directory.
+
+{{</Aside>}}
+
+Run the [`wrangler generate`](/workers/cli-wrangler/commands#generate) command to create a Worker using a [template](/workers/get-started/quickstarts#templates). Wrangler templates are git repositories that are designed to be a starting point for building a new Cloudflare Workers project. By default, the [default starter](https://github.com/cloudflare/worker-template) template will be used to generate your new Worker project:
+
+```sh
+wrangler generate <your-worker-name>
```
Comment (marciocloudflare) [inline content/r2/get-started/_index.md]:
```suggestion
wrangler generate <YOUR_WORKER_NAME>
```

File: content/r2/get-started/_index.md [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+```
+
+Next, find your newly generated `wrangler.toml` file in your project's directory and update `account_id` with your Cloudflare Account ID. 
+
+Find your Account ID by going logging in to the Cloudflare dashboard > **Overview** > move down to **API** > and select **Click to copy** to copy your **Account ID**. Or run the `wrangler whoami` command [to copy your Account ID](/workers/get-started/guide#6-preview-your-project).
```
Comment (marciocloudflare) [inline content/r2/get-started/_index.md]:
```suggestion
Find your Account ID by logging in to the Cloudflare dashboard > **Overview** > move down to **API** > and select **Click to copy** to copy your **Account ID**. Or run the `wrangler whoami` command [to copy your Account ID](/workers/get-started/guide#6-preview-your-project).
```

File: content/r2/get-started/_index.md [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+
+```toml
+name = "<your-worker-name>"
```
Comment (marciocloudflare) [inline content/r2/get-started/_index.md]:
```suggestion
name = "<YOUR_WORKER_NAME>"
```

File: content/r2/get-started/_index.md [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+type = "javascript"
+compatibility_date = "2022-02-10"
+ 
+account_id = "your-account-id" # ← Replace with your Account ID.
```
Comment (marciocloudflare) [inline content/r2/get-started/_index.md]:
```suggestion
account_id = "YOUR_ACCOUNT_ID" # ← Replace with your Account ID.
```

File: content/r2/get-started/_index.md [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+workers_dev = true
+```
+
+To bind your R2 bucket to your Worker, add the following to your `wrangler.toml` file. Update the `binding` property to a valid JavaScript variable identifier and `bucket_name` to the `<your-bucket-name>` you used to create your bucket in [step 3](#create-your-bucket):
```
Comment (marciocloudflare) [inline content/r2/get-started/_index.md]:
```suggestion
To bind your R2 bucket to your Worker, add the following to your `wrangler.toml` file. Update the `binding` property to a valid JavaScript variable identifier and `bucket_name` to the `<YOUR_BUCKET_NAME>` you used to create your bucket in [step 3](#create-your-bucket):
```

File: content/r2/get-started/_index.md [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+
+```toml
+[[r2_buckets]]
+binding = 'MY_BUCKET' # <~ valid JavaScript variable name
+bucket_name = '<your_bucket_name>'
```
Comment (marciocloudflare) [inline content/r2/get-started/_index.md]:
```suggestion
bucket_name = '<YOUR_BUCKET_NAME>'
```

File: content/r2/get-started/_index.md [hunk 14]
```diff
@@ -0,0 +1,268 @@
...
+```
+
+Find more detailed information on configuring your Worker in the [Wrangler Configuration documentation](/workers/cli-wrangler/configuration).
+
+## 5. Accessing your R2 bucket from your Worker
```
Comment (marciocloudflare) [inline content/r2/get-started/_index.md]:
```suggestion
## 5. Access your R2 bucket from your Worker
```

Comment (marciocloudflare) [reply content/r2/get-started/_index.md]: As per PCX style guide conventions

File: content/r2/get-started/_index.md [hunk 15]
```diff
@@ -0,0 +1,268 @@
...
+
+## 6. Bucket access and privacy
+
+With the above code added to your Worker, every incoming request has the ability to interact with your bucket. This means your bucket is publicly exposed and its contents can be accessed and modified by undesired actors. 
+
+You must now define authorization logic to determine who can perform what actions to your bucket. This logic lives within your Worker's code, as it is your application's job to determine user privileges. The following is a short list of resources related to access and authorization practices:
+
+1. [Basic Authentication](/workers/examples/basic-auth): Shows how to restrict access using the HTTP Basic schema.
+2. [Using Custom Headers](/workers/examples/auth-with-headers): Allow or deny a request based on a known pre-shared key in a header.
+3. [Authorizing users with Auth0](/workers/tutorials/authorize-users-with-auth0#overview): Integrate Auth0, an identity management platform, into a Cloudflare Workers application.
```
Comment (marciocloudflare) [inline content/r2/get-started/_index.md]:
```suggestion
1. [Basic Authentication](/workers/examples/basic-auth): Shows how to restrict access using the HTTP Basic schema.
1. [Using Custom Headers](/workers/examples/auth-with-headers): Allow or deny a request based on a known pre-shared key in a header.
1. [Authorizing users with Auth0](/workers/tutorials/authorize-users-with-auth0#overview): Integrate Auth0, an identity management platform, into a Cloudflare Workers application.
```

File: content/r2/get-started/_index.md [hunk 16]
```diff
@@ -0,0 +1,268 @@
...
+bucket_name = '<your_bucket_name>'
+```
+
+Find more detailed information on configuring your Worker in the [Wrangler Configuration documentation](/workers/cli-wrangler/configuration).
+
+## 5. Accessing your R2 bucket from your Worker
+
+Within your Worker code, your bucket is now available under the `MY_BUCKET` variable and you can begin interacting with it.
+
+An R2 bucket is able to READ, LIST, WRITE, and DELETE objects. You can see an example of all operations below using the Service Worker syntax. Add the following snippet into your project's `index.js` file:
```
Comment (marciocloudflare) [inline content/r2/get-started/_index.md]: I think READ, LIST, etc, should be between backticks. But as we don't have a guidance for this still, I'm leaving as it is. Something to discuss in content critique

File: content/r2/get-started/_index.md [hunk 17]
```diff
@@ -0,0 +1,268 @@
...
+
+```sh
+wrangler secret put AUTH_KEY_SECRET
+```
+
+This command will prompt you to enter a secret in your terminal:
+
+```sh
+wrangler secret put AUTH_KEY_SECRET
+Enter the secret text you'd like assigned to the variable AUTH_KEY_SECRET on the script named <your-worker-name>:
```
Comment (marciocloudflare) [inline content/r2/get-started/_index.md]:
```suggestion
Enter the secret text you'd like assigned to the variable AUTH_KEY_SECRET on the script named <YOUR_WORKER_NAME>:
```

File: content/r2/get-started/_index.md [hunk 18]
```diff
@@ -0,0 +1,268 @@
...
+*********
+🌀  Creating the secret for script name <your-worker-name>
```
Comment (marciocloudflare) [inline content/r2/get-started/_index.md]:
```suggestion
🌀  Creating the secret for script name <YOUR_WORKER_NAME>
```

File: content/r2/_index.md (hunk 2 above)
Comment (kristianfreeman) [reply content/r2/_index.md]: I think this is causing the page to render in the sidebar as a list, but it has no children - is that intended?

File: content/r2/_index.md [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+
+Cloudflare R2 Storage allows developers to store large amounts of unstructured data without the large egress bandwidth fees associated with typical cloud storage services.
+
+{{<Aside type="note" header="Beta registration">}}
+
+As of February 2022, R2 is currently under development. You can [sign up to join the waitlist for access](https://www.cloudflare.com/r2-storage/).
+
+{{</Aside>}}
+
+## Community
```
Comment (kristianfreeman) [inline content/r2/_index.md]: IMO since there's nothing R2-related here, we may want to remove this for now

File: content/r2/get-started/_index.md [hunk 20]
```diff
@@ -0,0 +1,268 @@
...
+$ curl https://your-worker.dev/cat-pic.jpg
+#=> test
+# Note: This is the value that was successfully PUT above
+```
+
+By completing this guide, you have successfully installed Wrangler and deployed your R2 bucket to Cloudflare.
+
+## Related resources
+
+[Workers Tutorials](https://developers.cloudflare.com/workers/tutorials)
```
Comment (kristianfreeman) [inline content/r2/get-started/_index.md]: This needs a newline after it for the links to render on separate lines

File: content/r2/get-started/_index.md (hunk 15 above)
Comment (lukeed) [reply content/r2/get-started/_index.md]: `1.` and `2.` and `3.` is preferred. The `format` step will rewrite all ordered lists to this format anyway.

File: content/r2/get-started/_index.md (hunk 20 above)
Comment (lukeed) [reply content/r2/get-started/_index.md]: Probably better to be a list anyway

File: package.json [hunk 21]
```diff
@@ -20,7 +20,7 @@
     "@types/prettier": "2.4.3",
     "@types/prismjs": "1.16.6",
     "prettier": "2.5.1",
-    "prismjs": "1.26.0",
+    "prismjs": "^1.27.0",
```
Comment (lukeed) [inline package.json]: Please revert this change

File: content/r2/get-started/_index.md (hunk 5 above)
Comment (deadlypants1973) [inline content/r2/get-started/_index.md]:
```suggestion
To install [`wrangler`](https://github.com/cloudflare/wrangler), ensure you have [`npm` installed](https://www.npmjs.com/get-npm). Use a Node version manager like [Volta](https://volta.sh/) or [nvm](https://github.com/nvm-sh/nvm) to avoid permission issues or to easily change Node.js versions, then run:
```

File: content/r2/get-started/_index.md (hunk 3 above)
Comment (deadlypants1973) [inline content/r2/get-started/_index.md]:
```suggestion
Cloudflare R2 Storage allows developers to store large amounts of unstructured data without the costly egress bandwidth fees associated with typical cloud storage services.
```

File: content/r2/get-started/_index.md [hunk 22]
```diff
@@ -0,0 +1,268 @@
...
+
+You must now define authorization logic to determine who can perform what actions to your bucket. This logic lives within your Worker's code, as it is your application's job to determine user privileges. The following is a short list of resources related to access and authorization practices:
+
+1. [Basic Authentication](/workers/examples/basic-auth): Shows how to restrict access using the HTTP Basic schema.
+2. [Using Custom Headers](/workers/examples/auth-with-headers): Allow or deny a request based on a known pre-shared key in a header.
+3. [Authorizing users with Auth0](/workers/tutorials/authorize-users-with-auth0#overview): Integrate Auth0, an identity management platform, into a Cloudflare Workers application.
+
+Continuing with your newly created bucket and Worker, you will need to protect all bucket operations. 
+
+For PUT and DELETE requests, you will make use of a new `AUTH_KEY_SECRET` environment variable, which you will define later as a Wrangler secret.
```
Comment (sdnts) [inline content/r2/get-started/_index.md]: Maybe it is worth mentioning that this in _one_ way to protect your bucket, but not the only way?

File: content/r2/get-started/_index.md [hunk 23]
```diff
@@ -0,0 +1,268 @@
...
+
+After running the `list` command, you will see all bucket names, including the one you have just created.
+
+## 4. Bind your bucket to a Worker
+
+You will need to bind your bucket to a Worker.
+
+{{<Aside type="note" header="Bindings">}}
+
+A binding is a runtime variable that the Workers runtime provides to your code. The capability of these variables can perform a number of operations depending on what they are bound to; for example, a binding may offer the [KV Namespace API](/workers/runtime-apis/kv), the [Durable Object API](/workers/runtime-apis/durable-objects), the [R2 API](#api), or it may be one of your user-defined environment variables. Every binding's variable name and behavior is determined by you when deploying the Worker. Refer to the [Environment Variables](/workers/platform/environment-variables) documentation for more information.
```
Comment (sdnts) [inline content/r2/get-started/_index.md]: Maybe another way to explain this could be:

A binding is how your Worker interacts with external resources such as [KV Namespaces], [Durable Objects], or [R2 Buckets]. You can declare a variable name in your `wrangler.toml` that will be bound to these resources at runtime, and you can interact with them by through this variable. Refer to the [Environment Variables](/workers/platform/environment-variables) documentation for more information.

**Review by sdnts — APPROVED** [review]

File: content/r2/get-started.md [hunk 24]
```diff
@@ -0,0 +1,264 @@
...
+3. [Authorizing users with Auth0](/workers/tutorials/authorize-users-with-auth0#overview): Integrate Auth0, an identity management platform, into a Cloudflare Workers application.
+
+Continuing with your newly created bucket and Worker, you will need to protect all bucket operations. 
+
+For PUT and DELETE requests, you will make use of a new `AUTH_KEY_SECRET` environment variable, which you will define later as a Wrangler secret. 
+
+For GET requests, you will ensure that only a specific file can be requested. All of this custom logic occurs inside of an `authorizeRequest` function, with the `hasValidHeader` function handling the custom header logic. If all validation passes, then the operation is allowed.
+
+```js
+const ALLOW_LIST = ['cat-pic.jpg'];
```
Comment (jculvey) [inline content/r2/get-started.md:179]: I would move this declaration down a few lines to right above `authorizeRequest`

File: content/r2/get-started.md [hunk 25]
```diff
@@ -0,0 +1,264 @@
...
+
+After running the `list` command, you will see all bucket names, including the one you have just created.
+
+## 4. Bind your bucket to a Worker
+
+You will need to bind your bucket to a Worker.
+
+{{<Aside type="note" header="Bindings">}}
+
+A binding is a runtime variable that the Workers runtime provides to your code. The capability of these variables can perform a number of operations depending on what they are bound to; for example, a binding may offer the [KV Namespace API](/workers/runtime-apis/kv), the [Durable Object API](/workers/runtime-apis/durable-objects), the [R2 API](#api), or it may be one of your user-defined environment variables. Every binding's variable name and behavior is determined by you when deploying the Worker. Refer to the [Environment Variables](/workers/platform/environment-variables) documentation for more information.
```
Comment (deadlypants1973) [inline content/r2/get-started.md]:
```suggestion
A binding is a how your Worker interacts with external resources such as [KV Namespaces](/workers/runtime-apis/kv), [Durable Objects](/workers/runtime-apis/durable-objects), or [R2 Buckets](#api). A binding is a runtime variable that the Workers runtime provides to your code. You can declare a variable name in your `wrangler.toml` file that will be bound to these resources at runtime, and interact with them through this variable. Every binding's variable name and behavior is determined by you when deploying the Worker. Refer to the [Environment Variables](https://github.com/workers/platform/environment-variables) documentation for more information.
```

**Review by marciocloudflare — APPROVED** [review]

File: content/r2/_index.md [hunk 26]
```diff
@@ -0,0 +1,22 @@
...
+
+{{<content-column>}}
+
+# Cloudflare R2 (beta) documentation
+
+Cloudflare R2 Storage allows developers to store large amounts of unstructured data without the costly egress bandwidth fees associated with typical cloud storage services.
+
+{{<Aside type="note" header="Beta registration">}}
+
+As of February 2022, R2 is currently under development. You can [sign up to join the waitlist for access](https://www.cloudflare.com/r2-storage/).
```
Comment (deadlypants1973) [inline content/r2/_index.md]:
```suggestion
As of February 2022, R2 is under development. You can [sign up to join the waitlist for access](https://www.cloudflare.com/r2-storage/).
```

File: content/r2/get-started.md [hunk 27]
```diff
@@ -0,0 +1,264 @@
...
+  title: Get started guide
+---
+
+# R2 (beta) get started guide
+
+Cloudflare R2 Storage allows developers to store large amounts of unstructured data without the costly egress bandwidth fees associated with typical cloud storage services.
+
+{{<Aside type="note" header="Beta registration">}}
+
+As of February 2022, R2 is currently under development. You can [sign up to join the waitlist for access](https://www.cloudflare.com/r2-storage/).
```
Comment (deadlypants1973) [inline content/r2/get-started.md]:
```suggestion
As of February 2022, R2 is under development. You can [sign up to join the waitlist for access](https://www.cloudflare.com/r2-storage/).
```

File: content/r2/get-started.md [hunk 28]
```diff
@@ -0,0 +1,264 @@
...
+```sh
+$ yarn global add @cloudflare/wrangler
+```
+or install with cargo:
+
+```sh
+$ cargo install wrangler
+```
+Refer to the Wrangler [Install/Update](/cli-wrangler/install-update) page for more information.
+
```
Comment (deadlypants1973) [inline content/r2/get-started.md:41]:
```suggestion
To install [`wrangler`](https://github.com/cloudflare/wrangler), ensure you have [`npm` installed](https://www.npmjs.com/get-npm). Use a Node version manager like [Volta](https://volta.sh/) or [nvm](https://github.com/nvm-sh/nvm) to avoid permission issues or to easily change Node.js versions, then run:

```sh
$ npm install -g @cloudflare/wrangler
```

or install with yarn:

```sh
$ yarn global add @cloudflare/wrangler
```

or install with cargo:

```sh
$ cargo install wrangler
```

Refer to the Wrangler [Install/Update](/cli-wrangler/install-update) page for more information.

```
